<?xml version="1.0"?>
<!--
    Launch default odometry and localization
-->

<launch>
    <!-- Launch file arguments -->

    <!-- Main argument -->
    <arg name="is_sim"                  default="false"/>
    <arg name="is_indoors"              default="true" />
    <arg name="initial_pose_x"          default="0.0" />
    <arg name="initial_pose_y"          default="0.0" />
    <arg name="initial_pose_a"          default="0.0" />

    <!-- low-level interface (serial) -->
    <arg name="start_serial"            default="false" /> 
    <!-- Should be set to true, if only the localization is running 
    (i.e. this is needed in order to have wheel encoders and actuation_to_twist working) -->
    <arg name="lli_port"                default="/dev/ttyACM0"/>

    <!-- map server -->
    <arg name="map_file"                default=""/>

    <!-- Lidar  -->
    <arg name="lidar_ip"                default="192.168.3.11" />

    <!-- Wheel encoder -->
    <arg name="start_wheel_encoders"    default="false" />

    <!-- GPS -->
    <arg name="start_rtk"               default="false"/>
    <arg name="gps_device"              default="/dev/ttyACM0" />
    <arg name="delay"                   default="2.0"/>
    <arg name="gps_baud"                default="250000" />  <!-- 38400 for UART connection -->
    <!-- yaw offest = initial_pose_a -->

    <!-- IMU -->
    <!-- Set to true if the imu is connected to tx2. Set to false if the imu is connected to teensy. -->
    <arg name="start_imu_on_tx2"        default="false"/>

    <!-- odom to map node -->
    <arg name="map_frame"               default="map"/>
    <arg name="wait_for_transform"      default="false"/>
    <arg name="publish_odometry"        default="true"/>
    <arg name="publish_pose"            default="true"/>

    <!-- Nodes -->

    <!-- Start low-level interface -->
    <group if="$(arg start_serial)">
        <node pkg="rosserial_python" type="serial_node.py" name="serial_node">
            <param name="port" value="$(arg lli_port)"/>
            <param name="baud" value="250000"/>
        </node>
    </group>

    <!-- Start map server -->
    <node if="$(eval map != '')" pkg="map_server" type="map_server" name="map_server" output="screen" args="$(arg map_file)"/>

    <!-- start all sensors if this is not a simulation -->
    <group unless="$(arg is_sim)">
        <param name="use_sim_time" value="false"/>
        <group if="$(arg is_indoors)">
            <!-- Start the Real Sense camera -->
            <include file="$(find svea_sensors)/launch/sensors_launch/real_sense_camera.launch"></include>

            <!-- Start Lidar -->
            <include file="$(find svea_sensors)/launch/sensors_launch/lidar.launch">
                <arg name="lidar_ip"                    value="$(arg lidar_ip)" />
            </include>

            <!-- Start AMCL -->
            <node pkg="amcl" type="amcl" name="amcl_localization" output="screen">
                <rosparam command="load"        file="$(find svea_sensors)/params/amcl/localize.yaml"/>
                <param name="scan"              value="scan"/>
                <param name="initial_pose_x"    value="$(arg initial_pose_x)" />
                <param name="initial_pose_y"    value="$(arg initial_pose_y)" />
                <param name="initial_pose_a"    value="$(arg initial_pose_a)" />
            </node>
        </group>

        <!-- Start GPS if outdoor localization -->
        <group unless="$(arg is_indoors)">
            <group if="$(arg start_rtk)">
                <include file="$(find svea_sensors)/launch/sensors_launch/rtk.launch">
                    <arg name="gps_device"      value="$(arg gps_device)" />
                    <arg name="baud"            value="$(arg gps_baud)" />
                </include>
            </group>
            <include file="$(find svea_sensors)/launch/sensors_launch/navsat.launch">
                <arg name="delay"           value="$(arg delay)"/>
                <arg name="yaw_offset"      value="$(arg initial_pose_a)"/>
            </include>
        </group>

        <!-- Start IMU -->
        <include if="$(arg start_imu_on_tx2)" file="$(find svea_sensors)/launch/sensors_launch/imu.launch"></include>

        <!-- Start controller -->
        <include file="$(find svea_sensors)/launch/sensors_launch/controller.launch"></include>

        <!-- Start wheel encoder -->
        <include if="$(arg start_wheel_encoders)" file="$(find svea_sensors)/launch/sensors_launch/wheel_encoder.launch"></include>

        <!-- Static transforms for sensors -->
        <include file="$(find svea_sensors)/launch/transforms.launch">
            <arg name="is_indoors"  value="$(arg is_indoors)"/>
        </include>
    </group>

    <node pkg="svea_sensors" type="odom_to_map" name="odom_to_map" output="screen">
        <param name="map_frame"             value="$(arg map_frame)"/>
        <param name="wait_for_transform"    value="$(arg wait_for_transform)"/>
        <param name="publish_odometry"      value="$(arg publish_odometry)"/>
        <param name="publish_pose"          value="$(arg publish_pose)"/>
    </node>

    <!-- Do not start the sensors, useful when running from a bag file -->
    <group if="$(arg is_sim)">
        <param name="use_sim_time"          value="true" />
    </group>

    <!-- Robot localization (Local EKF)-->
    <node pkg="robot_localization" type="ekf_localization_node" name="ekf_local" clear_params="true" output="screen">
        <rosparam command="load" file="$(find svea_sensors)/params/robot_localization/local_ekf.yaml" />
        <!-- initial_state: [x,          y,          z, 
                             roll,       pitch,      yaw, 
                             x_dot,      y_dot,      z_dot, 
                             roll_dot,   pitch_dot,  yaw_dot, 
                             x_dot_dot,  y_dot_dot,  z_dot_dot]-->
        <rosparam subst_value="true">
        initial_state: [$(arg initial_pose_x),  $(arg initial_pose_y),  0.0, 
                        0.0,                    0.0,                    $(arg initial_pose_a), 
                        0.0,                    0.0,                    0.0, 
                        0.0,                    0.0,                    0.0, 
                        0.0,                    0.0,                    0.0]
        </rosparam>
        <param name="publish_tf"            value="true"/>
        <param name="odom0"                 value="rs/t265_camera/odom/sample"/>
        <param name="imu0"                  value="imu/data"/>
        <param name="twist0"                value="actuation_twist"/>
        <param name="twist1"                value="wheel_encoder_twist"/>
    </node>

    <!-- Global EKF -->
    <node unless="$(arg is_indoors)" pkg="robot_localization" type="ekf_localization_node" name="ekf_global" clear_params="true" output="screen">
        <rosparam command="load" file="$(find svea_sensors)/params/robot_localization/global_ekf.yaml"/>
        <!-- initial_state: [x,          y,          z, 
                             roll,       pitch,      yaw, 
                             x_dot,      y_dot,      z_dot, 
                             roll_dot,   pitch_dot,  yaw_dot, 
                             x_dot_dot,  y_dot_dot,  z_dot_dot]-->
        <rosparam subst_value="true">
        initial_state: [$(arg initial_pose_x),  $(arg initial_pose_y),  0.0, 
                        0.0,                    0.0,                    $(arg initial_pose_a), 
                        0.0,                    0.0,                    0.0, 
                        0.0,                    0.0,                    0.0, 
                        0.0,                    0.0,                    0.0]
        </rosparam>
        <param name="publish_tf"            value="true"/>
        <param name="odom0"                 value="odometry/gps"/>
        <param name="imu0"                  value="imu/data"/>
        <param name="twist0"                value="actuation_twist"/>
        <param name="twist1"                value="wheel_encoder_twist"/>
        <remap from="/odometry/filtered"    to="/odometry/filtered/global"/>
        <remap from="/set_pose"             to="/global/set_pose"/>
    </node>
</launch>
